From 54c3dfa96da62576da8ebb56dd57a66f8b5bfd2c Mon Sep 17 00:00:00 2001
From: Christophe Priouzeau <christophe.priouzeau@st.com>
Date: Mon, 21 May 2018 14:26:47 +0200
Subject: [PATCH 10/14] waylandsink: Uprank to secondary

Signed-off-by: Christophe Priouzeau <christophe.priouzeau@st.com>
---
 ext/wayland/gstwaylandsink.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/ext/wayland/gstwaylandsink.c b/ext/wayland/gstwaylandsink.c
index 739dd44..dff8724 100644
--- a/ext/wayland/gstwaylandsink.c
+++ b/ext/wayland/gstwaylandsink.c
@@ -1167,7 +1167,7 @@ plugin_init (GstPlugin * plugin)
 
   gst_wl_shm_allocator_register ();
 
-  return gst_element_register (plugin, "waylandsink", GST_RANK_MARGINAL,
+  return gst_element_register (plugin, "waylandsink", GST_RANK_SECONDARY,
       GST_TYPE_WAYLAND_SINK);
 }
 
-- 
2.7.4

