From 678d91a743bb35e202211abc6db495fbf8d75601 Mon Sep 17 00:00:00 2001
From: Christophe Priouzeau <christophe.priouzeau@st.com>
Date: Tue, 18 Feb 2020 13:33:21 +0100
Subject: [PATCH 14/14] waylandsink add waylandpool on meson build

Signed-off-by: Christophe Priouzeau <christophe.priouzeau@st.com>
---
 ext/wayland/meson.build | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/ext/wayland/meson.build b/ext/wayland/meson.build
index a3ffb70..66dd148 100644
--- a/ext/wayland/meson.build
+++ b/ext/wayland/meson.build
@@ -5,7 +5,8 @@ wl_sources = [
     'wldisplay.c',
     'wlwindow.c',
     'wlvideoformat.c',
-    'wllinuxdmabuf.c'
+    'wllinuxdmabuf.c',
+    'waylandpool.c'
 ]
 
 libdrm_dep = dependency('libdrm', version: '>= 2.4.55', required:get_option('wayland'))
-- 
2.25.1

