From fd4ce2545ebf30263de41684e7b7c68e6fd1e3be Mon Sep 17 00:00:00 2001
From: Christophe Priouzeau <christophe.priouzeau@st.com>
Date: Fri, 14 Feb 2020 15:25:05 +0100
Subject: [PATCH] authorize pulse audio request via dbus

Signed-off-by: Christophe Priouzeau <christophe.priouzeau@st.com>
---
 src/bluetooth.conf | 17 +++++++++++++++++
 1 file changed, 17 insertions(+)

diff --git a/src/bluetooth.conf b/src/bluetooth.conf
index 8a1e258..380388d 100644
--- a/src/bluetooth.conf
+++ b/src/bluetooth.conf
@@ -22,6 +22,23 @@
     <allow send_interface="org.mpris.MediaPlayer2.Player"/>
   </policy>
 
+  <policy user="pulse">
+    <allow own="org.bluez"/>
+    <allow send_destination="org.bluez"/>
+    <allow send_interface="org.bluez.Agent1"/>
+    <allow send_interface="org.bluez.MediaEndpoint1"/>
+    <allow send_interface="org.bluez.MediaPlayer1"/>
+    <allow send_interface="org.bluez.ThermometerWatcher1"/>
+    <allow send_interface="org.bluez.AlertAgent1"/>
+    <allow send_interface="org.bluez.Profile1"/>
+    <allow send_interface="org.bluez.HeartRateWatcher1"/>
+    <allow send_interface="org.bluez.CyclingSpeedWatcher1"/>
+    <allow send_interface="org.bluez.GattCharacteristic1"/>
+    <allow send_interface="org.bluez.GattDescriptor1"/>
+    <allow send_interface="org.freedesktop.DBus.ObjectManager"/>
+    <allow send_interface="org.freedesktop.DBus.Properties"/>
+  </policy>
+
   <policy context="default">
     <allow send_destination="org.bluez"/>
   </policy>
-- 
2.17.1

